// @appbaseio/reactivesearch/lib/components/basic/ReactiveComponent.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _actions = require('@appbaseio/reactivecore/lib/actions');

var _helper = require('@appbaseio/reactivecore/lib/utils/helper');

var _types = require('@appbaseio/reactivecore/lib/utils/types');

var _types2 = _interopRequireDefault(_types);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ReactiveComponent = function (_Component) {
	_inherits(ReactiveComponent, _Component);

	function ReactiveComponent(props) {
		_classCallCheck(this, ReactiveComponent);

		var _this = _possibleConstructorReturn(this, _Component.call(this, props));

		_initialiseProps.call(_this);

		_this.internalComponent = null;

		var _props$onQueryChange = props.onQueryChange,
		    onQueryChange = _props$onQueryChange === undefined ? null : _props$onQueryChange;


		_this.setQuery = function (obj) {
			_this.props.updateQuery(_extends({}, obj, {
				componentId: props.componentId,
				label: props.filterLabel,
				showFilter: props.showFilter,
				onQueryChange: onQueryChange,
				URLParams: props.URLParams
			}));
		};

		if (props.defaultQuery) {
			_this.internalComponent = props.componentId + '__internal';
		}
		return _this;
	}

	ReactiveComponent.prototype.componentWillMount = function componentWillMount() {
		this.props.addComponent(this.props.componentId);
		if (this.internalComponent) {
			this.props.addComponent(this.internalComponent);
		}

		this.setReact(this.props);

		// set query for internal component
		if (this.internalComponent && this.props.defaultQuery) {
			var _props$defaultQuery = this.props.defaultQuery(),
			    query = _props$defaultQuery.query,
			    queryOptions = _objectWithoutProperties(_props$defaultQuery, ['query']);

			if (queryOptions) {
				this.props.setQueryOptions(this.internalComponent, queryOptions, false);
			}

			this.props.updateQuery({
				componentId: this.internalComponent,
				query: query || null
			});
		}
	};

	ReactiveComponent.prototype.componentWillUnmount = function componentWillUnmount() {
		this.props.removeComponent(this.props.componentId);

		if (this.internalComponent) {
			this.props.removeComponent(this.internalComponent);
		}
	};

	ReactiveComponent.prototype.render = function render() {
		var _this2 = this;

		var _props = this.props,
		    children = _props.children,
		    addFn = _props.addComponent,
		    watchFn = _props.watchComponent,
		    removeFn = _props.removeComponent,
		    queryOptionsFn = _props.setQueryOptions,
		    updateFn = _props.updateQuery,
		    rest = _objectWithoutProperties(_props, ['children', 'addComponent', 'watchComponent', 'removeComponent', 'setQueryOptions', 'updateQuery']);

		try {
			var childrenWithProps = _react2.default.Children.map(children, function (child) {
				return _react2.default.cloneElement(child, _extends({}, rest, { setQuery: _this2.setQuery }));
			});
			return _react2.default.createElement(
				'div',
				null,
				childrenWithProps
			);
		} catch (e) {
			return null;
		}
	};

	return ReactiveComponent;
}(_react.Component);

var _initialiseProps = function _initialiseProps() {
	var _this3 = this;

	this.setReact = function (props) {
		var react = props.react;


		if (react) {
			if (_this3.internalComponent) {
				var newReact = (0, _helper.pushToAndClause)(react, _this3.internalComponent);
				props.watchComponent(props.componentId, newReact);
			} else {
				props.watchComponent(props.componentId, react);
			}
		} else if (_this3.internalComponent) {
			props.watchComponent(props.componentId, { and: _this3.internalComponent });
		}
	};
};

ReactiveComponent.defaultProps = {
	URLParams: false,
	showFilter: true
};

ReactiveComponent.propTypes = {
	componentId: _types2.default.stringRequired,
	hits: _types2.default.data,
	aggregations: _types2.default.selectedValues,
	selectedValue: _types2.default.selectedValue,
	addComponent: _types2.default.funcRequired,
	removeComponent: _types2.default.funcRequired,
	watchComponent: _types2.default.funcRequired,
	updateQuery: _types2.default.funcRequired,
	setQueryOptions: _types2.default.funcRequired,
	onQueryChange: _types2.default.func,
	URLParams: _types2.default.boolRequired,
	showFilter: _types2.default.bool,
	filterLabel: _types2.default.string,
	defaultQuery: _types2.default.func,
	react: _types2.default.react,
	children: _types2.default.children
};

var mapStateToProps = function mapStateToProps(state, props) {
	return {
		hits: state.hits[props.componentId] && state.hits[props.componentId].hits || [],
		aggregations: state.aggregations[props.componentId] && state.aggregations[props.componentId] || null,
		selectedValue: state.selectedValues[props.componentId] && state.selectedValues[props.componentId].value || null
	};
};

var mapDispatchtoProps = function mapDispatchtoProps(dispatch) {
	return {
		addComponent: function addComponent(component) {
			return dispatch((0, _actions.addComponent)(component));
		},
		removeComponent: function removeComponent(component) {
			return dispatch((0, _actions.removeComponent)(component));
		},
		watchComponent: function watchComponent(component, react) {
			return dispatch((0, _actions.watchComponent)(component, react));
		},
		updateQuery: function updateQuery(updateQueryObject) {
			return dispatch((0, _actions.updateQuery)(updateQueryObject));
		},
		setQueryOptions: function setQueryOptions(component, props, execute) {
			return dispatch((0, _actions.setQueryOptions)(component, props, execute));
		}
	};
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchtoProps)(ReactiveComponent);
