// @appbaseio/reactivesearch/lib/components/basic/TextField.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _actions = require('@appbaseio/reactivecore/lib/actions');

var _helper = require('@appbaseio/reactivecore/lib/utils/helper');

var _types = require('@appbaseio/reactivecore/lib/utils/types');

var _types2 = _interopRequireDefault(_types);

var _Input = require('../../styles/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Title = require('../../styles/Title');

var _Title2 = _interopRequireDefault(_Title);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TextField = function (_Component) {
	_inherits(TextField, _Component);

	function TextField(props) {
		_classCallCheck(this, TextField);

		var _this = _possibleConstructorReturn(this, _Component.call(this, props));

		_initialiseProps.call(_this);

		_this.type = 'match';
		_this.state = {
			currentValue: ''
		};
		_this.locked = false;
		return _this;
	}

	TextField.prototype.componentWillMount = function componentWillMount() {
		this.props.addComponent(this.props.componentId);
		this.setReact(this.props);

		if (this.props.selectedValue) {
			this.setValue(this.props.selectedValue, true);
		} else if (this.props.defaultSelected) {
			this.setValue(this.props.defaultSelected, true);
		}
	};

	TextField.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
		var _this2 = this;

		(0, _helper.checkPropChange)(this.props.react, nextProps.react, function () {
			_this2.setReact(nextProps);
		});

		(0, _helper.checkPropChange)(this.props.dataField, nextProps.dataField, function () {
			_this2.updateQuery(_this2.state.currentValue, nextProps);
		});

		if (this.props.defaultSelected !== nextProps.defaultSelected) {
			this.setValue(nextProps.defaultSelected, true, nextProps);
		} else if (
		// since, selectedValue will be updated when currentValue changes,
		// we must only check for the changes introduced by
		// clear action from SelectedFilters component in which case,
		// the currentValue will never match the updated selectedValue
		this.props.selectedValue !== nextProps.selectedValue && this.state.currentValue !== nextProps.selectedValue) {
			this.setValue(nextProps.selectedValue || '', true, nextProps);
		}
	};

	TextField.prototype.componentWillUnmount = function componentWillUnmount() {
		this.props.removeComponent(this.props.componentId);
	};

	TextField.prototype.setReact = function setReact(props) {
		if (props.react) {
			props.watchComponent(props.componentId, props.react);
		}
	};

	TextField.prototype.render = function render() {
		return _react2.default.createElement(
			'div',
			{ style: this.props.style, className: this.props.className },
			this.props.title && _react2.default.createElement(
				_Title2.default,
				{ className: (0, _helper.getClassName)(this.props.innerClass, 'title') || null },
				this.props.title
			),
			_react2.default.createElement(_Input2.default, {
				type: 'text',
				className: (0, _helper.getClassName)(this.props.innerClass, 'input') || null,
				placeholder: this.props.placeholder,
				onChange: this.handleChange,
				value: this.state.currentValue
			})
		);
	};

	return TextField;
}(_react.Component);

var _initialiseProps = function _initialiseProps() {
	var _this3 = this;

	this.defaultQuery = function (value, props) {
		if (value && value.trim() !== '') {
			var _this3$type, _ref;

			return _ref = {}, _ref[_this3.type] = (_this3$type = {}, _this3$type[props.dataField] = value, _this3$type), _ref;
		}
		return null;
	};

	this.handleTextChange = (0, _helper.debounce)(function (value) {
		_this3.updateQuery(value, _this3.props);
	}, this.props.debounce);

	this.setValue = function (value) {
		var isDefaultValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
		var props = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _this3.props;

		// ignore state updates when component is locked
		if (props.beforeValueChange && _this3.locked) {
			return;
		}

		_this3.locked = true;
		var performUpdate = function performUpdate() {
			_this3.setState({
				currentValue: value
			}, function () {
				if (isDefaultValue) {
					_this3.updateQuery(value, props);
				} else {
					// debounce for handling text while typing
					_this3.handleTextChange(value);
				}
				_this3.locked = false;
			});
		};
		(0, _helper.checkValueChange)(props.componentId, value, props.beforeValueChange, props.onValueChange, performUpdate);
	};

	this.updateQuery = function (value, props) {
		var query = props.customQuery || _this3.defaultQuery;

		var _props$onQueryChange = props.onQueryChange,
		    onQueryChange = _props$onQueryChange === undefined ? null : _props$onQueryChange;


		props.updateQuery({
			componentId: props.componentId,
			query: query(value, props),
			value: value,
			label: props.filterLabel,
			showFilter: props.showFilter,
			onQueryChange: onQueryChange,
			URLParams: props.URLParams
		});
	};

	this.handleChange = function (e) {
		_this3.setValue(e.target.value);
	};
};

TextField.propTypes = {
	addComponent: _types2.default.funcRequired,
	componentId: _types2.default.stringRequired,
	defaultSelected: _types2.default.string,
	react: _types2.default.react,
	removeComponent: _types2.default.funcRequired,
	dataField: _types2.default.stringRequired,
	title: _types2.default.title,
	beforeValueChange: _types2.default.func,
	onValueChange: _types2.default.func,
	customQuery: _types2.default.func,
	onQueryChange: _types2.default.func,
	updateQuery: _types2.default.funcRequired,
	placeholder: _types2.default.string,
	selectedValue: _types2.default.selectedValue,
	filterLabel: _types2.default.string,
	URLParams: _types2.default.boolRequired,
	showFilter: _types2.default.bool,
	style: _types2.default.style,
	className: _types2.default.string,
	innerClass: _types2.default.style,
	debounce: _types2.default.number
};

TextField.defaultProps = {
	placeholder: 'Search',
	URLParams: false,
	showFilter: true,
	style: {},
	className: null,
	debounce: 0
};

var mapStateToProps = function mapStateToProps(state, props) {
	return {
		selectedValue: state.selectedValues[props.componentId] && state.selectedValues[props.componentId].value || null
	};
};

var mapDispatchtoProps = function mapDispatchtoProps(dispatch) {
	return {
		addComponent: function addComponent(component) {
			return dispatch((0, _actions.addComponent)(component));
		},
		removeComponent: function removeComponent(component) {
			return dispatch((0, _actions.removeComponent)(component));
		},
		watchComponent: function watchComponent(component, react) {
			return dispatch((0, _actions.watchComponent)(component, react));
		},
		updateQuery: function updateQuery(updateQueryObject) {
			return dispatch((0, _actions.updateQuery)(updateQueryObject));
		}
	};
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchtoProps)(TextField);
